### DISCUSSION TITLE
[mobile] Any way to run rust code in background ?

### DISCUSSION THREAD

**Discussion by Vexcited** [tauri-apps/tauri#11688] [Q&A]
I want to do polling on HTTP requests and trigger a notification when something has changed.
So I tried doing this that way :

```rust
#[tauri::command]
fn start_background_task(app: AppHandle) {
    tauri::async_runtime::spawn(async move {
        loop {
            let timestamp = chrono::Local::now().to_rfc3339();
            let message = format!("background task running at {}", timestamp);

            app.emit("background-task", &message).unwrap();
            println!("background task ran at {}", timestamp);
            
            // sleep for 1 minute
            std::thread::sleep(std::time::Duration::from_secs(60));
        }
    });
}
```

and I listened to events in the JS code and I did `sendNotification` in the front-end like this :

```javascript
async function run_task() {
  const response = await fetch("https://...");
  const json = await response.json() as {
    id: string
    // ...
  };

  sendNotification({
    title: "My App Name",
    body: `Updated for ${json.id}`,
    visibility: Visibility.Public
  });
}

listen<void>("background-task", (event) => {
  console.info("background task received; id:", event.id);
  run_task();
})
```

However, I'm not receiving any notification when I close the app (not killing it but just closing it) while I do receive them when the app is currently active.

---

I know I can add Kotlin code to make a worker that could poll and create the notification in background, but that would make me rewrite the JS code I have in Kotlin and probably Swift for iOS.

The intention here is to have the exact same code and make it work everywhere...

---

Any idea ? Maybe this is a feature idea ?
Also FCM is not possible in this case so the new `push-notification` plugin that is currently in progress will probably be no use in there.

Thanks !

**Comment by probablykasper** [comment]
Afaik mobile is really restrictive about running code in the background. You might need to use some OS-provided API to schedule code to run in the background (and even that has been a bit hit-or-miss when I tried it in Flutter)

**Reply by nicholasdgoodman** [reply]
I have a hunch that this would require a Mobile Plugin with matching Kotlin / Swift code; but I would hope that in future releases of Tauri this would be one of those APIs provided "out of the box" so to speak.

A good insight into what is required can be gleaned from looking at the similar documentation for the Background Runner feature in Capacitor.js:

https://capacitorjs.com/docs/apis/background-runner
